lookout/style/format/analyzer.py

```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
+        """
         cls.log.info("train %s %s %s", ptr.url, ptr.commit, data)
         files = data["files"]
+        language_to_files = defaultdict(list)
         for file in files:
+            if not len(file.uast.children):
+                continue
+            # TODO: replace this temporary hack when langauge fields in File objects are fixed
+            extension = file.path.rsplit(".", 1)[1]
+            if extension != "js":
+                continue
+            language_to_files["javascript"].append(file)
             cls.log.info("%s %s %d", file.path, file.language, len(file.uast.children))
 
-        """
-        Plan:
-
-        1. Load all the File-s - they are streamed and we need to read them all into a list
-        2. Sort by language
-        3. Loop per language group:
-        4. extract_features()
-        5. train the base model - tree or forest
-        6. train the rules
-        7. generate the Modelforge model
-        """
+        model = FormatModel().construct(cls, ptr)
 
-        return FormatModel().construct(cls, ptr)
+        for language, files in language_to_files.items():
+            cls.log.info("processing %s files", language)
+            cls.log.info("extracting features from %d files", len(files))
+            fe = FeatureExtractor(language=language, siblings_window=5, parents_depth=2)
+            X, y = fe.extract_features(files)
+            if X.shape[0] < cls.LOWER_BOUND_INSTANCES:
+                cls.log.warning("did not train a model for language %s: not sufficiently many "
+                                "instances (%d)" % (language, X.shape[0]))
+            cls.log.info("training rules model")
+            bscv = BayesSearchCV(
+                TrainableRules(prune_branches_algorithms=["reduced-error"], prune_attributes=False,
+                               random_state=42),
+                {"base_model_type_string": Categorical(["sklearn.ensemble.RandomForestClassifier",
+                                                        "sklearn.tree.DecisionTreeClassifier"]),
+                 "max_depth": Categorical([None, 5, 10]),
+                 "max_features": Categorical([None, "auto"]),
+                 "min_samples_split": Integer(2, 20),
+                 "min_samples_leaf": Integer(1, 20)},
+                n_jobs=-1,
+                n_iter=30)
+            bscv.fit(X, y)
+            cls.log.info("score of the best estimator found: %.3f", bscv.best_score_)
+            cls.log.info("params of the best estimator found: %s", str(bscv.best_params_))
+            cls.log.info("training the model with complete data")
+            trainable_rules = TrainableRules(prune_branches_algorithms=["reduced-error"],
+                                             prune_attributes=True, random_state=42,
+                                             uncertain_attributes=True, **bscv.best_params_)
+            trainable_rules.fit(X, y)
+            cls.log.info("saving %s rules model", language)
```

Comment:
we don't really do anything special - take the `rules` attribute and put it into the model. can be removed.